# Welcome to Lab: Illini 🧡💙

Over the past week, you have learned about lists and functions in Python, and how to add new columns to a DataFrame!  You'll dive into using real-world applications of these topics in lab this week, while exploring data about the over 1,300 football games the Fighting Illini have played over the past 130 years of football history! 

A few tips to remember:

- **You are not alone on your journey in learning programming!**  You have your lab TA, your CAs, your lab group, and the professors (Prof. Wade and Prof. Karle), who are all here to help you out!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

In [1]:
# Meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = "KunlunZhu"
ca1_name = "Caroline"
ca2_name = "Emily"
ca3_name = ""

# Say hello to each other!
# - Groups of 3 are ideal :)
# - However groups of 2 or 4 are fine too!
partner1_name = "Xinmeng Zhang"
partner1_netid = "xinmeng8"

partner2_name = ""
partner2_netid = ""

partner3_name = ""
partner3_netid = ""

# Which lab section are you in? Record it below! It should start with a Y and have 
# 2 numbers after the Y. If you're unsure - ask your TA/CAs!
lab_section = "Y23"

<hr style="color: #DD3403;">

# Part 1: Illini Football

This year has been one of the best years for Illini Football, and we'll use our skills as Data Scientists to create the ultimate Illini Football DataFrame!

To start off, we'll use the "Illini Football" dataset you were introduced during the first week of lecture. The dataset is available here:
```
http://waf.cs.illinois.edu/discovery/football.csv
```

Load this dataset into a DataFrame called `df`!

In [2]:
import pandas as pd
df=pd.read_csv("http://waf.cs.illinois.edu/discovery/football.csv")
df

,Season,Date,Location,Opponent,OpponentRank,Result,IlliniScore,OpponentScore,Note
0,2025,2025-09-13,vs.,Western Michigan,NaN,W,38.0,0.0,NaN
1,2025,2025-09-06,@,Duke,NaN,W,45.0,19.0,NaN
2,2025,2025-08-29,vs.,Western Illinois,NaN,W,52.0,3.0,NaN
3,2024,2024-12-31,"Orlando, FL",South Carolina,14.0,W,21.0,17.0,Cheez-It Citrus Bowl
4,2024,2024-11-30,@,Northwestern,NaN,W,38.0,28.0,NaN
...,...,...,...,...,...,...,...,...,...
1307,1892,1892-10-24,@,Nebraska,NaN,L,0.0,6.0,NaN
1308,1892,1892-10-22,"Omaha, NE",Doane,NaN,W,20.0,0.0,NaN
1309,1892,1892-10-21,@,Washington,NaN,W,22.0,0.0,NaN
1310,1892,1892-10-12,vs.,Northwestern,NaN,T,16.0,16.0,NaN


### Part 1.1: Calculate the Total Score

In a game of football, two teams play against each other to score points.  Add a new column to the DataFrame called `TotalScore` that adds the `IlliniScore` and `OpponentScore` together:

In [3]:
# Add a `TotalScore` column to the DataFrame:
...
df["TotalScore"]=df["IlliniScore"]+df["OpponentScore"]

### Part 1.2: The Craziest Game of Illini Football

Let's consider the "craziest" game of Illini Football the game where the **most total points** were scored.

Create a DataFrame `df_craziest` that contains just one row of the very craziest game:

In [4]:
df_craziest=df.nlargest(1, 'TotalScore')

### Part 1.3: Find the Worst Game in Illini Football

Let's consider the "worst" game to be the game where the Illini lost by the most points (this is not the game where the opponent scored the most points, but the game where their score had the biggest difference). Create a DataFrame `df_worst` that contains the row(s) with the greatest score difference.

You may want to create a new column to store the "spread" between the the games!
There might be more than one row with the same score difference, so you need to add the use `keep='all'` as an parameter to `df.nlargest()` or `df.nsmallest()` to not drop the duplicates.


In [5]:
df['spread']=df['OpponentScore']-df['IlliniScore']

df_worst=df.nlargest(1,'spread',keep='all')
df_worst

,Season,Date,Location,Opponent,OpponentRank,Result,IlliniScore,OpponentScore,Note,TotalScore,spread
76,2018,2018-11-17,vs.,Iowa,NaN,L,0.0,63.0,NaN,63.0,63.0
1171,1906,1906-11-17,@,Chicago,NaN,L,0.0,63.0,NaN,63.0,63.0


### Part 1.4: The Best Homecoming Game

In football, the "homecoming game" is a football game played at the team's home stadium and is often accompanied with celebrations all week ("Homecoming Week").  In this dataset, homecoming games are denoted by the `"Note"` column containing the string `"Homecoming"`.

Find the "best" homecoming game! In other words, find the homecoming game where the Illini won by the most points. Create a DataFrame `df_homecoming_best` that contains just one row of the best homecoming game:

In [6]:
df_homecoming = df[df['Note'] == 'Homecoming']
df_homecoming_best=df_homecoming.nlargest(1,'IlliniScore')

In [7]:
### TEST CASE for Part 1: The Best and Worst Games in History
tada = "\N{PARTY POPPER}"

assert( 'df' in vars() ), "You appear to not have the correct dataset loaded in `df`."
assert( len(df) > 1300 ), "You appear to not have the correct dataset loaded in `df`."
assert( "IlliniScore" in df ), "You appear to not have the correct dataset loaded in `df`."
assert( "OpponentScore" in df ), "You appear to not have the correct dataset loaded in `df`."

assert( 'df_worst' in vars() ), "df_worst is not defined."
assert( len(df_worst) == 2 ), "df_worst appears incorrect."
assert( 'df_craziest' in vars() ), "df_craziest is not defined."
assert( len(df_craziest) == 1 ), "df_craziest appears incorrect."
assert( 'df_homecoming_best' in vars() ), "df_homecoming_best is not defined."
assert( len(df_homecoming_best) == 1 ), "df_homecoming_best appears incorrect."

OS = list(df.columns).index("OpponentScore")
IS = list(df.columns).index("IlliniScore")
__tdf = df.dropna(subset=["IlliniScore"])
__tdfh = __tdf[__tdf.Note == "Homecoming"]
assert( df_worst.iloc[0].OpponentScore == __tdf.loc[(__tdf.iloc[:, OS] - __tdf.iloc[:, IS]).sort_values().index].iloc[-1, OS] ), "Puzzle #1 appears incorrect."
assert( df_craziest.iloc[0].OpponentScore == __tdf.loc[(__tdf.iloc[:, OS] + __tdf.iloc[:, IS]).sort_values().index].iloc[-1, OS] ), "Puzzle #1 appears incorrect."
assert( df_homecoming_best.iloc[0].IlliniScore == __tdfh.loc[(__tdfh.iloc[:, OS] - __tdfh.iloc[:, IS]).sort_values().index].iloc[0, IS] ), "Puzzle #1 appears incorrect."
assert( df_homecoming_best.iloc[0].OpponentScore == __tdfh.loc[(__tdfh.iloc[:, OS] - __tdfh.iloc[:, IS]).sort_values().index].iloc[0, OS] ), "Puzzle #1 appears incorrect."
print(f"{tada} All Tests Passed! {tada}")


🎉 All Tests Passed! 🎉


<hr style="color: #DD3403;">

## Part 2: Using Functions for Calculating Values

One of the most useful ways to use functions in Python is to **apply** the function to every row of a DataFrame to generate a new value.  You will write functions of increasing complexity and apply them to the Illini football dataset:

### Part 2.1: Create the Simplest Function

Write a function called `three` that accepts one parameter called `row` and **always** returns `-3`.  You should be able to do this in just 2 lines of code:

In [8]:
# Write the function three:
def three(row):
    return -3

To **apply** a function across every row in your DataFrame, you will use the syntax:
```py
df.apply( <function name ONLY>, axis=1 )
```

For example, the following code applies the function `three` to all of the rows in our Illini Football DataFrame:
```
df.apply(three, axis=1)
```

This is the ONLY time in this course we will not put parentheses after the function name!

Run the code below to see the result of `three` on every index in your DataFrame:

In [9]:
df.apply(three, axis=1)

0      -3
1      -3
2      -3
3      -3
4      -3
       ..
1307   -3
1308   -3
1309   -3
1310   -3
1311   -3
Length: 1312, dtype: int64

We want to save this value into our DataFrame, so we'll create a new column named `score` that will store the result of the apply operation.  We've provided you this code for this first one.  When you run the code:
- Note that there's a new column named `score` on the far-right hand side.
- Note that the value is always `-3`, since that's the return value of your function.

In [10]:
# Creates a new column named `score` that will store the results
# of applying the `three` function across every single row individually:
df["score"] = df.apply(three, axis=1)
df

,Season,Date,Location,Opponent,OpponentRank,Result,IlliniScore,OpponentScore,Note,TotalScore,spread,score
0,2025,2025-09-13,vs.,Western Michigan,NaN,W,38.0,0.0,NaN,38.0,-38.0,-3
1,2025,2025-09-06,@,Duke,NaN,W,45.0,19.0,NaN,64.0,-26.0,-3
2,2025,2025-08-29,vs.,Western Illinois,NaN,W,52.0,3.0,NaN,55.0,-49.0,-3
3,2024,2024-12-31,"Orlando, FL",South Carolina,14.0,W,21.0,17.0,Cheez-It Citrus Bowl,38.0,-4.0,-3
4,2024,2024-11-30,@,Northwestern,NaN,W,38.0,28.0,NaN,66.0,-10.0,-3
...,...,...,...,...,...,...,...,...,...,...,...,...
1307,1892,1892-10-24,@,Nebraska,NaN,L,0.0,6.0,NaN,6.0,6.0,-3
1308,1892,1892-10-22,"Omaha, NE",Doane,NaN,W,20.0,0.0,NaN,20.0,-20.0,-3
1309,1892,1892-10-21,@,Washington,NaN,W,22.0,0.0,NaN,22.0,-22.0,-3
1310,1892,1892-10-12,vs.,Northwestern,NaN,T,16.0,16.0,NaN,32.0,0.0,-3


In [11]:
### TEST CASE for Part 2.1: Using df.apply
tada = "\N{PARTY POPPER}"
from inspect import signature

assert( "three" in vars() ), "You appear to not have the a function named `three`."
assert( type(three).__name__ == "function" ), "You appear to not have the a function named `three`."
assert( len(signature(three).parameters) == 1 ), "You function `three` must take exactly one parameter."
assert( three(None) == -3 ), "You function `three` must always return -3."
assert( "score" in df ), "You must have a `score` column in your DataFrame `df`."
assert( df["score"].max() == -3 ), "The value of `score` must be -3 for every row."
assert( df["score"].min() == -3 ), "The value of `score` must be -3 for every row."
print(f"{tada} All Tests Passed! {tada}")


🎉 All Tests Passed! 🎉


### Part 2.2: Using the `row` parameter

In your function, you had one parameter named `row`.  When calling a function through `apply`, the current row that is being evaluated by the `apply` function will be sent as the `row`.  This will allow you to have access to the column values for the row just like you access a column in a DataFrame.

For example, when inside of your function, you can do the following:
- `row["Opponent"]` will contain the name of the opponent (ex: `"Purdue"`, `"Nebraska"`, etc)
- `row["IlliniScore"]` will contain the score of the Illini (ex: `45`, `23`, etc)
- ...and so on, for all the columns.

Write a function named `calculateTotalScore` that accepts one parameter called `row` and **always** returns the total score of the row by adding the `IlliniScore` value to the `OpponentScore` value.  You should be able to this with one line to define the function and one line as a return statement:

In [12]:
# Write the function calculateTotalScore:
def calculateTotalScore(row):
    return row["IlliniScore"] + row["OpponentScore"]

Now, create a new column named `CalculatedTotalScore` that stores the result of applying the `calculateTotalScore` function to the DataFrame:

In [13]:
# Creates a new column named `CalculatedTotalScore` that will store the results
# of applying the `calculateTotalScore` function across every single row individually:
...
df["CalculatedTotalScore"] = df.apply(calculateTotalScore, axis=1)
df


,Season,Date,Location,Opponent,OpponentRank,Result,IlliniScore,OpponentScore,Note,TotalScore,spread,score,CalculatedTotalScore
0,2025,2025-09-13,vs.,Western Michigan,NaN,W,38.0,0.0,NaN,38.0,-38.0,-3,38.0
1,2025,2025-09-06,@,Duke,NaN,W,45.0,19.0,NaN,64.0,-26.0,-3,64.0
2,2025,2025-08-29,vs.,Western Illinois,NaN,W,52.0,3.0,NaN,55.0,-49.0,-3,55.0
3,2024,2024-12-31,"Orlando, FL",South Carolina,14.0,W,21.0,17.0,Cheez-It Citrus Bowl,38.0,-4.0,-3,38.0
4,2024,2024-11-30,@,Northwestern,NaN,W,38.0,28.0,NaN,66.0,-10.0,-3,66.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
1307,1892,1892-10-24,@,Nebraska,NaN,L,0.0,6.0,NaN,6.0,6.0,-3,6.0
1308,1892,1892-10-22,"Omaha, NE",Doane,NaN,W,20.0,0.0,NaN,20.0,-20.0,-3,20.0
1309,1892,1892-10-21,@,Washington,NaN,W,22.0,0.0,NaN,22.0,-22.0,-3,22.0
1310,1892,1892-10-12,vs.,Northwestern,NaN,T,16.0,16.0,NaN,32.0,0.0,-3,32.0


In [14]:
### TEST CASE for Part 2.2: Using the `row` parameter
tada = "\N{PARTY POPPER}"
from inspect import signature
assert( "calculateTotalScore" in vars() ), "You appear to not have the a function named `calculateTotalScore`."
assert( type(calculateTotalScore).__name__ == "function" ), "You appear to not have the a function named `calculateTotalScore`."
assert( len(signature(calculateTotalScore).parameters) == 1 ), "You function `calculateTotalScore` must take exactly one parameter."
assert( calculateTotalScore(df.loc[0]) == df.loc[0].IlliniScore + df.loc[0].OpponentScore ), "You function `calculateTotalScore` did not calculate the total score."
assert( calculateTotalScore(df.loc[100]) == df.loc[100].IlliniScore + df.loc[100].OpponentScore ), "You function `calculateTotalScore` did not calculate the total score."
assert( "CalculatedTotalScore" in df ), "You must have a `CalculatedTotalScore` column in your DataFrame `df`."
assert( df["CalculatedTotalScore"].sum() - (df.IlliniScore.sum() + df.OpponentScore.sum()) == 0 ), "The value of `CalculatedTotalScore` must be the total score."
print(f"{tada} All Tests Passed! {tada}")


🎉 All Tests Passed! 🎉


### Part 2.3: Labeling the Game as "Home" or "Away"

Functions become extremely useful when we begin to introduce conditionals to process data based on the value of the row.  In the function below, we are using a new concept -- an `if` statement in Python -- to return if the game is a "Home" game played in Champaign, IL.

The contents inside of the `if` statement are a conditional, just like you used in a DataFrame! 🎉

Run the cell below to define the `homeOrAway` function:

In [15]:
# (You do not need to modify this function, but you MUST run it to define this function.)

def homeOrAway(row):
  # We first check if the data in the `Location` row is "vs.".
  # ...if it is "vs.", we know the game was a "Home" game:
  if row["Location"] == "vs.": return "Home"

  # If we did not return in the previous line of Python code,
  # we know the game was not a "Away" game:
  return "Away"

Create a new column named `GameType` that store the result of applying the `homeOrAway` function to the DataFrame:

In [16]:
# Creates a new column named `GameType` that will store the results
# of applying the `homeOrAway` function across every single row individually:
df["GameType"]=df.apply(homeOrAway, axis=1)

### 👀 Verify Your DataFrame! 👀

There is no automated test cases for this puzzle or the rest of this lab!  You should inspect your DataFrame to ensure your code produced the expected result before moving onto the next section! :)

<hr style="color: #DD3403;">

## Part 3: Finding the Win Rate for a Team

During your lab section, your TA will give your group a specific opponent to focus on.  Write the name of the opponent you're focusing on here:

*(If you were not in lab section, check the #announcement channel on Discord for the team you should work on.)*

In [17]:
opponentFocus ='Penn State'
opponentFocus

'Penn State'

### Part 3.1: Calculate the "Win Rate" of the Illini Against Your Opponent

Calculate the percentage of times the Illini won a football game when playing against the team you're investigating (ties do not count).  To do this, make sure you are considering **ONLY** the games played against the team you're investigating.

*You may need multiple cells to do this, add more "Code" cells as necessary.*

In [18]:
# Calculate the "Win Rate" vs. Opponent:
# (You may need multiple cells to do this, add more "Code" cells as necessary.)
Opponent_name='Penn State'
df_vs_opponent = df[df["Opponent"] == Opponent_name]
df_no_ties=df_vs_opponent[df_vs_opponent["IlliniScore"]!=df_vs_opponent["OpponentScore"]]

wins=(df_no_ties['IlliniScore']>=df_no_ties['OpponentScore']).sum()
total_games = df_no_ties.shape[0]

win_rate = (wins / total_games) * 100
win_rate

np.float64(21.428571428571427)

### Part 3.2: Calculate the "Win Rate" of the Illini at Home Games

Still focusing ONLY on the opponent you're investigating, calculate the win rate of "Home Games" against the team you're investigating.  To do this, you now want to only consider the set of games played in Champaign, IL with the Opponent being the team you're investigating.

*You may need multiple cells to do this, add more "Code" cells as necessary.*

In [19]:
# Calculate the "Win Rate" of Home Games vs. Opponent:
# (You may need multiple cells to do this, add more "Code" cells as necessary.)
Opponent_name='Penn State'
dfHomeVsOpponent=df[(df["Opponent"] == Opponent_name) &(df["Location"] == "vs.")]
df

dfHomeNoTies = dfHomeVsOpponent[dfHomeVsOpponent["IlliniScore"] != dfHomeVsOpponent["OpponentScore"]]
homeWins = (dfHomeNoTies["IlliniScore"] > dfHomeNoTies["OpponentScore"]).sum()
totalHomeGames=dfHomeNoTies.shape[0]
if totalHomeGames > 0:
    homewinrate = (homeWins / totalHomeGames) * 100
else:
    homewinrate = 0 

homewinrate

np.float64(28.57142857142857)

In [20]:
Opponent_name='Penn State'
dfHomeVsOpponent = df[(df["Opponent"] == Opponent_name) & (df["Location"] == "Champaign,Il")]
df



,Season,Date,Location,Opponent,OpponentRank,Result,IlliniScore,OpponentScore,Note,TotalScore,spread,score,CalculatedTotalScore,GameType
0,2025,2025-09-13,vs.,Western Michigan,NaN,W,38.0,0.0,NaN,38.0,-38.0,-3,38.0,Home
1,2025,2025-09-06,@,Duke,NaN,W,45.0,19.0,NaN,64.0,-26.0,-3,64.0,Away
2,2025,2025-08-29,vs.,Western Illinois,NaN,W,52.0,3.0,NaN,55.0,-49.0,-3,55.0,Home
3,2024,2024-12-31,"Orlando, FL",South Carolina,14.0,W,21.0,17.0,Cheez-It Citrus Bowl,38.0,-4.0,-3,38.0,Away
4,2024,2024-11-30,@,Northwestern,NaN,W,38.0,28.0,NaN,66.0,-10.0,-3,66.0,Away
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1307,1892,1892-10-24,@,Nebraska,NaN,L,0.0,6.0,NaN,6.0,6.0,-3,6.0,Away
1308,1892,1892-10-22,"Omaha, NE",Doane,NaN,W,20.0,0.0,NaN,20.0,-20.0,-3,20.0,Away
1309,1892,1892-10-21,@,Washington,NaN,W,22.0,0.0,NaN,22.0,-22.0,-3,22.0,Away
1310,1892,1892-10-12,vs.,Northwestern,NaN,T,16.0,16.0,NaN,32.0,0.0,-3,32.0,Home


### Part 3.3: 👀 Verify Your Results! 👀

**You are about to share your analysis with others.**  Double-check that you understand the code you wrote and that it answers the questions.  Other groups are going to rely on your data for analysis, and everyone has different teams, so it's important to double-check your work!

### Part 3.4: Complete Your Illini Opponent Win Rate Table

Edit the next cell to do the following:
- Add the result of **your group's analysis to the first cell**
- **Talk to other groups around you** to find two other opponents and record their analysis in the second and third rows
- *(You can have more than three rows, but a minimum of three is required.  If you are NOT in lab section, check Discord #labs channel for some class-wide values that will be posted on Friday.)*

(**Double-Click to Edit this Cell** -- Replace contents of cells with "✏️" with data!)

| Opponent | Overall Win Rate vs. Opponent | Win Rate at Home Games vs. Opponent |
| -------- | ----------------------------- | ----------------------------------- |
| Penn State |21.428571428571427 | 28.57142857142857|
|  Penn State| 21.428571428571427| 28.57142857142857 |
|
| ✏️Opponent #3✏️ | ✏️Overall Win Rate ✏️ | ✏️Win Rate at Home✏️ |




<hr style="color: #DD3403;">

## Part 4: Analysis

In the next cell, answer the following questions about your analysis:

How does the Illini performance compare between the opponents in your summary table?  Is one opponent easier/harder than the others?  Your answer **MUST** include the use of specific opponent names from your table above, use specific data values (actual numbers), and be one or more complete sentences.

After answering the first question, write at least one more complete sentence explaining if you believe the Illini do better at home games or away games.  Explain why you came to that conclusion **using specific opponent names and data values from your table above**.

In my analysis, the total win rate of 21.4% so the Illini have a relatively low overall win rate.

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)